homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,232 @@
+"""The Subaru integration."""
+import asyncio
+from datetime import datetime, timedelta
+import logging
+import time
+
+from subarulink import (
+    Controller as SubaruAPI,
+    InvalidPIN,
+    PINLockoutProtect,
+    SubaruException,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_DEVICE_ID,
+    CONF_PASSWORD,
+    CONF_PIN,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    API_GEN_2,
+    CONF_HARD_POLL_INTERVAL,
+    COORDINATOR_NAME,
+    DEFAULT_HARD_POLL_INTERVAL,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    ENTRY_CONTROLLER,
+    ENTRY_COORDINATOR,
+    ENTRY_LISTENER,
+    ENTRY_VEHICLES,
+    SUPPORTED_PLATFORMS,
+    VEHICLE_API_GEN,
+    VEHICLE_HAS_EV,
+    VEHICLE_HAS_REMOTE_SERVICE,
+    VEHICLE_HAS_REMOTE_START,
+    VEHICLE_HAS_SAFETY_SERVICE,
+    VEHICLE_LAST_UPDATE,
+    VEHICLE_NAME,
+    VEHICLE_VIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass, base_config):
+    """Set up of Subaru component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Subaru from a config entry."""
+    config = entry.data
+    websession = aiohttp_client.async_get_clientsession(hass)
+    date = datetime.now().strftime("%Y-%m-%d")
+    device_name = "Home Assistant: Added " + date
+    try:
+        controller = SubaruAPI(
+            websession,
+            config[CONF_USERNAME],
+            config[CONF_PASSWORD],
+            config[CONF_DEVICE_ID],
+            config[CONF_PIN],
+            device_name,
+            update_interval=entry.options.get(
+                CONF_HARD_POLL_INTERVAL, DEFAULT_HARD_POLL_INTERVAL
+            ),
+        )
+        if not await controller.connect():
+            _LOGGER.error("Failed to connect")
+            raise ConfigEntryNotReady
+    except SubaruException as ex:
+        _LOGGER.error("Unable to communicate with Subaru API: %s", ex.message)
+        raise ConfigEntryNotReady from ex
+
+    vehicle_info = {}
+    for vin in controller.get_vehicles():
+        vehicle_info[vin] = get_vehicle_info(controller, vin)
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            return await subaru_update(vehicle_info, controller)
+        except InvalidPIN as err:
+            _LOGGER.error(
```

Comment:
You shouldn't have your own error logs, since the updater takes care of only printing once instead of filling the logs with them.
I would also change to `raise UpdateFailed(err) from err`, otherwise the updater log won't contain any specific information.

Reply:
Changes made.